File: huffman.js
string="malayalam";


//find the frequency of the characters in string
function frequency(string)
{
	freq={};
	for(count=0; count<string.length; count++)
	{
		if (string[count] in freq)
			freq[string[count]]+=1;
		else
			freq[string[count]]=1;
	}
	return freq;
}
freqs=frequency(string);
//freqs=frequency("aaabccdeeeeeffg");
//{'a': 3, 'c': 2, 'b': 1, 'e': 5, 'd': 1, 'g': 1, 'f': 2}



// frequency sorted and stored order of rank
function sortFreq(freqs)
{
	tuples=[];
	for(var key in freqs)
		//console.log(freqs[key],key);
		tuples.push([freqs[key],key]);	
	return tuples.sort();
}
tuples=sortFreq(freqs);
//[[1, 'b'], [1, 'd'], [1, 'g'], [2, 'c'], [2, 'f'], [3, 'a'], [5, 'e']] 



//build tree
function buildTree(tuples)
{
	while(tuples.length>1)
	{
	leastTwo=[tuples[0][1],tuples[1][1]]
	//console.log(leastTwo);
	theRest=tuples.slice(2,tuples.length);
	//console.log(theRest);
	combFreq=tuples[0][0]+tuples[1][0];
	//console.log(combFreq);
	tuples=theRest;
	end=[combFreq,leastTwo];
	tuples.push(end);
	//console.log(tuples);
	tuples.sort();
	//console.log(tuples);
	}
	return tuples[0][1];
}
tree=buildTree(tuples);	
//console.log(tree);
//[ 'a', [ 'm', [ 'y', 'l' ] ] ]


code={};
pat='';
//assiging codes to each letter
function assignCode(node,pat)
{
	if(typeof(node)==typeof(""))
		code[node]=pat;
	else
	{
		assignCode( node[0], pat+'0');
		assignCode( node[1], pat+'1');
	}
}
assignCode(tree,pat);
//console.log(code);
//{ a: '0', m: '10', y: '110', l: '111' }


//encoding given string
function encode(string)
{
	output='';
	for(s in string)
		output+=code[string[s]];
	return output;
}

encoded=encode(string);
console.log("Encoded as:",encoded);
//10011101100111010



//decoding that string
function decode(tree,encoded)
{
	output='';
	p=tree;
	for(bit in encoded)
	{
		if(encoded[bit]=='0')
			p=p[0];
		else
			p=p[1];
		if(typeof(p)==typeof(''))
		{
		output+=p;
		p=tree;
		}
	}
	return output;
}
decoded=decode(tree,encoded);
console.log("Decoded as:",decoded);
//malayalam

File: huffman.py
codes={}

#dictionary of words with frequency
def frequency(string):
	freq={}
	for word in string:
		freq[word]=freq.get(word,0)+1
	return freq

#print frequency("aaabccdeeeeeffg")   
#{'a': 3, 'c': 2, 'b': 1, 'e': 5, 'd': 1, 'g': 1, 'f': 2}


#sort the words in the order of ranking and stored in tuples
def sortFreq(freq):
	words=freq.keys()
	tuples=[]
	for wd in words:
		tuples.append((freq[wd],wd))
	tuples.sort()
	return tuples
	 
#print sortFreq(frequency("aaabccdeeeeeffg")) 
#[(1, 'b'), (1, 'd'), (1, 'g'), (2, 'c'), (2, 'f'), (3, 'a'), (5, 'e')]


#make a tree
def buildTree(tuples) :
	while len(tuples) > 1 :
		leastTwo = tuple(tuples[0:2])
		#print leastTwo
		theRest = tuples[2:]
		#print theRest
		combFreq = leastTwo[0][0] + leastTwo[1][0]
		#print combFreq
		tuples= theRest + [(combFreq,leastTwo)]
		#print tuples
		tuples.sort()
		#print len(tuples)
	return tuples[0]
#print buildTree(sortFreq(frequency("aaabccdeeeeeffg")))
#(15, ((6, ((3, 'a'), (3, ((1, 'g'), (2, 'c'))))), (9, ((4, ((2, 'f'), (2, ((1, 'b'), (1, 'd'))))), (5, 'e')))))


#trim left,right and combine
def trimTree(tree):
	p=tree[1]
	#print 'p',p
	if type(p)==type(''): 
		return p
	else:
		#print 'p[01]',p[0],p[1]
		return (trimTree(p[0]),trimTree(p[1]))

#print trimTree(buildTree(sortFreq(frequency("aaabccdeeeeeffg")))) 
#(('a', ('g', 'c')), (('f', ('b', 'd')), 'e'))


#assigning codes to all nodes
def assignCodes(nodes, pat=''):
	global codes
	#print nodes,pat
	if type(nodes)==type(''):
		codes[nodes]=pat
	else:
		assignCodes(nodes[0], pat+'0')
		assignCodes(nodes[1], pat+'1')
tree=trimTree(buildTree(sortFreq(frequency("aaabccdeeeeeffg"))))
assignCodes(tree)
print codes 
#{'a': '00', 'c': '011', 'b': '1010', 'e': '11', 'd': '1011', 'g': '010', 'f': '100'}



#original string encoded with the data
def encode( string):
	global codes
	output=''
	for ch in string:
		output+=codes[ch]
	return output
print encode("aaabccdeeeeeffg")
#000000101001101110111111111111100100010


#decode the data and create tree
def decode(tree, s):
	output=''
	p=tree
	for b in s:
		if b=='0':
			p=p[0]
		else:
			p=p[1]
		#print b,p
		if type(p)==type(''):
			output+=p
			p=tree
	return output
print decode(tree,'000000101001101110111111111111100100010')	
